### PR TITLE
Fix tests

### PR DESCRIPTION
**Pull request by tushar-rishav**
- [x] Modify tests to handle CLI args appropriately. Handles `py.test` cli args
otherwise it interferes with arg parser.
- [x] Configure travis to run py.test
- [x] Add more tests.

CC @Uran198

### CODE REVIEW

File: tests/treeTest.py [hunk 1]
```diff
@@ -9,11 +10,16 @@
...
+        self.old_argv = sys.argv
         log_printer = ListLogPrinter()
         file_dict = run_coala(log_printer=log_printer,
-                              autoapply=False)[2]
+                              autoapply=False,
+                              arg_list=[])[2]
         self.FL = list(parse_file_dict(file_dict).keys())
 
+    def tearDown(self):
+        sys.argv = self.old_argv
```
Comment (Uran198) [inline tests/treeTest.py]: You don't need `old_argv`, at least you're not using it.

**Comment by tushar-rishav** [comment]
@Uran198 16b8b71beb5bfaa661 looks good? I am comparing the 'tests' subtree completely

File: tests/treeTest.py [hunk 2]
```diff
@@ -1,18 +1,15 @@
+import os
 import unittest
-
```
Comment (Uran198) [inline tests/treeTest.py]: It's not big deal, but generally it looks nicer when you separate imports into blocks. If you really want to remove this line, make sure that you remove in all files, and make it in separate commit, as it only relates to code style.

File: tests/treeTest.py [hunk 3]
```diff
@@ -1,18 +1,15 @@
...
 
     def setUp(self):
-        log_printer = ListLogPrinter()
-        file_dict = run_coala(log_printer=log_printer,
-                              autoapply=False)[2]
-        self.FL = list(parse_file_dict(file_dict).keys())
+        self.FL = []
+        for root, _, file_list in os.walk(os.path.abspath('tests')):
+            for fname in file_list:
+                self.FL.append(os.path.join(root, fname))
```
Comment (Uran198) [inline tests/treeTest.py]: Here will be more preferable to use, `self.FL.extend(file_list)`. The amortized time complexity will be the same, but easier to read.

**Comment by Uran198** [comment]
LGTM 16b8b71, other then some small minor comments.
What do you think about, maybe, trying to get travis working with Python in this PR? It may be better to split, though.

**Comment by Uran198** [comment]
ack 44debb246b6988457567283fdb98b186c779685a

**Comment by Uran198** [comment]
Was there some issue with tests before 77ebcc583dff08eda608477cefd0153acdf28cff? Basically, why you do it?

**Comment by tushar-rishav** [comment]
@Uran198 Yeah previously passing any flag to `py.test` was throwing `coala: error: unrecognized arguments`. So I think it happens while [collecting](https://github.com/coala-analyzer/coala-html/commit/77ebcc583dff08eda608477cefd0153acdf28cff#diff-2111a24b47129eff676979e35ec2cf1dR59) the cli arguments, there are some extra flags that just sneak in while running tests for eg. `py.test --cov -k <sometest>`. Here coala arg_parser doesn't recognise `-k` and `--cov`. I can't think of any elegant solution other than overriding the `sys.argv`. Thoughts?

File: tests/coalaHTMLTest.py [hunk 4]
```diff
@@ -27,22 +29,28 @@ def test_output_file(self):
         update_file = ""
         noupdate_file = ""
         with prepare_file(["#todo this is todo"], None) as (lines, filename):
-            execute_coala(coala_html.main,
-                          "coala-html",
-                          "-c", os.devnull,
-                          "-b", "LineCountBear",
-                          "-f", re.escape(filename),
-                          "--nolaunch")
+            p_update = Process(target=execute_coala, args=(coala_html.main,
```
Comment (tushar-rishav) [inline tests/coalaHTMLTest.py]: If we agree on not covering tests for spawned server then I will revert back to old approach (not using Process)

File: tests/treeTest.py (hunk 3 above)
Comment (tushar-rishav) [inline tests/treeTest.py]: Oh. Sorry didn't notice this before. `file_list` just contains the name of the files and not the absolute paths. I need to use root to create the complete path.

**Comment by Uran198** [comment]
I may be too picky, but line length limit on the body of commit message is 72 chars, so this looks nicer:
```
While collecting the CLI argument list from `default_arg_parser` it
expects to have a valid entry points (sys.argv[0]). The entry points
might not be available in case of testing the wrapper around default arg
parser. Hence, it's essential to mock this behaviour.
```
for 42e3dd44a956a4c83d94358127cbd5578cc76e02. The same for other commits.

File: tests/treeTest.py (hunk 3 above)
Comment (Uran198) [inline tests/treeTest.py]: My points was to use `extend` instead of `apply`, as it's slightly faster. So you can do:
```python
self.FL.extend(os.path.join(root, fname) for fname in file_list)
```

File: tests/treeTest.py [hunk 5]
```diff
@@ -25,5 +22,6 @@ def test_extract(self):
 
     def test_graph(self):
         tr = Tree(self.FL)
-        testChild = tr.root.child['tests'].child.get('specs', None)
-        self.assertTrue(testChild)
+        # use `set` to ignore the  order in comparison.
+        testChild = set(tr.root.child.keys())
+        self.assertEqual(testChild, set(os.listdir('tests')))
```
Comment (Uran198) [inline tests/treeTest.py]: Here you can use [assertCountEqual](https://docs.python.org/3.2/library/unittest.html#unittest.TestCase.assertCountEqual) instead of converting to sets.

File: tests/treeTest.py (hunk 2 above)
Comment (Uran198) [inline tests/treeTest.py]: So what's about this, @tushar-rishav ?

**Comment by Uran198** [comment]
ack 9ca966008c1e0b31b5078757f19eacacac385e1a

File: coalahtml/helper.py [hunk 6]
```diff
@@ -51,7 +51,7 @@ def create_dir(path: str):
     if not os.path.exists(path):
         try:
             os.mkdir(path, mode=0o775)
-        except OSError as e:
+        except OSError as e:  # pragma: no cover
```
Comment (Uran198) [inline coalahtml/helper.py]: Why?
Can put this into commit message.

File: coalahtml/coala_html.py [hunk 7]
```diff
@@ -35,7 +35,7 @@ def main():
 
     dir_path = create_dir(os.path.abspath(args.dir))
 
-    if len(list(os.walk(dir_path))) > 0:
+    if len(list(os.walk(dir_path))) > 0:  # pragma: no cover
```
Comment (Uran198) [inline coalahtml/coala_html.py]: Basically, here you're testing if it's a directory or a file, right? If yes, then use respective methods, like `os.is_file` or `os.is_dir`, instead.

Also, why not cover this test case?

File: coalahtml/coala_html.py [hunk 8]
```diff
@@ -75,8 +75,8 @@ def main():
         # Launch server with reference point dir_path
         os.chdir(dir_path)
         if not os.path.exists('bower_components'):
-            res = call(['bower', 'install'])
-            if res != 0:
+            res = subprocess.call(['bower', 'install'])
+            if res != 0:  # pragma: no cover
```
Comment (Uran198) [inline coalahtml/coala_html.py]: Why not cover this?

File: tests/coalaHTMLTest.py [hunk 9]
```diff
@@ -16,13 +19,23 @@
 class coalaHTMLTest(unittest.TestCase):
 
     def setUp(self):
+        os.chdir(os.path.dirname(os.path.dirname(__file__)))
```
Comment (Uran198) [inline tests/coalaHTMLTest.py]: Why two times `os.path.dirname`?

File: tests/coalaHTMLTest.py [hunk 10]
```diff
@@ -40,10 +54,14 @@ def test_output_file(self):
...
-                          "--noupdate",
-                          "--nolaunch")
+                          "--noupdate")
 
             with open(self.result_file, 'r') as fp:
                 noupdate_file = fp.read()
 
         self.assertEqual(update_file, noupdate_file)
+        # We `chdir` before spawning the server
+        shutil.rmtree('../coalahtmlapp', ignore_errors=True)
```
Comment (Uran198) [inline tests/coalaHTMLTest.py]: It doesn't look like you're doing chdir here, also why not put all the cleanup into the tearDown?

**Comment by Uran198** [comment]
In 5b7749552f6c3ebf52e0f8c3fa36fb427d30a4e3 the changes aren't really related, at least it's not clear from the commit message why. IIUC you can mock `call` method with changing import style. Also, when you have to you "and" in commit message it's a good sign that this should be two separate commits.

**Comment by Uran198** [comment]
74838d89f1e3357ef247159b9e789fc535d6d95e: should go before 5b7749552f6c3ebf52e0f8c3fa36fb427d30a4e3, cause in later one you're changing directories in the same fashion, so it's kind of logically depends on it.
Also, why do you need to change the directory at all?

File: tests/treeTest.py [hunk 11]
```diff
@@ -7,6 +7,7 @@ class TreeTest(unittest.TestCase):
 
     def setUp(self):
         self.FL = []
+        os.chdir(os.path.dirname(os.path.dirname(__file__)))
```
Comment (Uran198) [inline tests/treeTest.py]: `os.path.dirname` is two time here, as well.

**Comment by Uran198** [comment]
Apart from what I wrote, some commit messages are not good. Please take a look, for example, at http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html or http://chris.beams.io/posts/git-commit/
Basically, think of them as continuing the sentence,"If applied, this commit will ..."

File: tests/coalaHTMLTest.py (hunk 10 above)
Comment (tushar-rishav) [inline tests/coalaHTMLTest.py]: No. I do `chdir` just before spawning the server in `coala_html.py` at [L76](https://github.com/coala-analyzer/coala-html/blob/tushar-rishav/coalahtml/coala_html.py#L76) :sweat_smile: 
Um we needn't put all cleanup into tearDown because the `rmtree` is required only when coala_html is run (as it creates necessary directories). Other test suites might not need to run coala_html. So In such case the clean up isn't required. For now the clean up is required only for this particular suite. Maybe we put all clean ups to tearDown when they are needed in other suites and leave for now?

File: tests/coalaHTMLTest.py (hunk 9 above)
Comment (tushar-rishav) [inline tests/coalaHTMLTest.py]: Um I just wanted to go two step back. Maybe there exists better approach? `os.path.abspath("../..")` maybe? I am not sure. :smile:

File: coalahtml/coala_html.py (hunk 7 above)
Comment (tushar-rishav) [inline coalahtml/coala_html.py]: Not exactly! I wanted to test if the target directory already has the angular files (copied). But it seems like I overlooked something. For example, consider an empty directory `tests`, the output from `list(os.walk('test'))` is `[('test', [], [])]` which is strange. Thoughts? I assumed it to be 0. So I believe this is a bug!
Now, I have two ways to fix this:
* Still use os.walk approach but change the count to 3 maybe?
* Just check if `dir_path/index.html` exists. This is efficient and neat. Fixing this in separate commit?
@Uran198

Comment (tushar-rishav) [inline coalahtml/coala_html.py]: ^^ Fixed and `pragma` removed

Comment (Uran198) [inline coalahtml/coala_html.py]: >Just check if `dir_path/index.html` exists.
This feels like a little too hardcoded. If you want to check if you didn't copy it already, why not check exactly that, that every file from the path you're copying is here already. This way if you're for example updating coala-html, e. g. adding new files, user will get those new files copied. With your approach, he will have to manually remove this directory.

File: tests/coalaHTMLTest.py (hunk 9 above)
Comment (Uran198) [inline tests/coalaHTMLTest.py]: Now, why do you wan to do this in the first place?

File: tests/coalaHTMLTest.py (hunk 10 above)
Comment (Uran198) [inline tests/coalaHTMLTest.py]: Either way, comment is not very informative.
Basically, all the tests in this class will benefit from this tearDown. :P
Also, I though that all the tests in this class, if there will be more, will be related to running coala_html, so they will require this cleanup.

File: tests/coalaHTMLTest.py [hunk 12]
```diff
@@ -40,10 +54,14 @@ def test_output_file(self):
                           "-c", os.devnull,
```
Comment (Uran198) [inline tests/coalaHTMLTest.py]: It feels wrong to have running coala_html.main two times, but testing stuff once. Probably, this should be two separate test cases.

**Comment by Uran198** [comment]
ack 5206c9c4597e55534cbd72f71d61d022b0aacb79 6689f1c37b3267ab4f2d8f7c08fa279746db3f1f

File: coalahtml/helper.py (hunk 6 above)
Comment (Uran198) [inline coalahtml/helper.py]: If we failed to make directory, does is make sense to continue execution?
If no, don't except or reraise, exception. Let if print the traceback (some may consider this ugly, but IMO the user will understand what happened from it, so it's enough) or handle exception at higher level and print user message.

**Comment by Uran198** [comment]
line wrapping is not at 72 characters:
37826d2521afd8eff0a4957a98acac40aaaf115e
c9201b455c7a1af8cf3a06d254f91c8d0b7dc979

**Comment by Uran198** [comment]
c9201b455c7a1af8cf3a06d254f91c8d0b7dc979 is not needed, IIRC, you can mock both ways.

File: coalahtml/coala_html.py (hunk 7 above)
Comment (Uran198) [inline coalahtml/coala_html.py]: And, why not cover?

File: coalahtml/helper.py (hunk 6 above)
Comment (tushar-rishav) [inline coalahtml/helper.py]: Hm makes sense, you are right! I will remove the exception catching here then. :+1:

File: coalahtml/coala_html.py (hunk 7 above)
Comment (tushar-rishav) [inline coalahtml/coala_html.py]: @Uran198 I agree. So the ideal solution is to match the contents of `coalahtml/_coalahtml/` and the `<target directory>`. IIUC I should better match the directory contents - filenames, file contents etc. I believe http://stackoverflow.com/a/24860799/3673031 is what I need. Thoughts? Is there any better approach?

Comment (tushar-rishav) [inline coalahtml/coala_html.py]: ^ Anyway this is going to be in https://github.com/coala-analyzer/coala-html/pull/81/commits/58f9d34bb324d1885b618d53dc406dd22ad221d4

Comment (Uran198) [inline coalahtml/coala_html.py]: Yes, sure.
>Anyway this is going to be in 58f9d34

As I see, there are two alternatives, one is that you only check that the file structure is the same. Then, maybe, it'll be easier for people to hack on stuff, if they wanted to. Another approach can be to check the file content as well. This way, if you'll be changing files in the new version you'll update them, but it's more costly and will give no hacking ability.

Comment (tushar-rishav) [inline coalahtml/coala_html.py]: Yes. Comparing file contents won't let user create a fork of the angular app and just keep the json configs. But that would mean they can't use the latest changes in angular app. What do we do?

**Comment by tushar-rishav** [comment]
reack fef43776252ba43b4c6fe9c6f39a564d3aff61a3

**Comment by Uran198** [comment]
ack 4edb1b67eb987c94a638a63d4707f5d98ecc6da8

**Comment by Uran198** [comment]
ack 5996cd677c72febaab5222ee1e595a15cca43395

File: coalahtml/coala_html.py (hunk 7 above)
Comment (Uran198) [inline coalahtml/coala_html.py]: I'll be happy with either decision. Evaluate all pros and cons and try to figure out the best trade off. There can be another solution to this. For example, asking user if we detect them some of the files are not what they should be if he wants to leave them as is, or update it. Maybe, make it a cli option for coala-html.

File: tests/coalaHTMLTest.py [hunk 13]
```diff
@@ -21,8 +24,33 @@ def setUp(self):
...
+                         mock_server_close,
+                         mock_serve_forever):
+        """
+        The approach is to call ``coala_html.main`` twice - first with
+        nolaunch and then with noupdate args. With only nolaunch being
+        passed, coala_html.main would generate the json configs and next
+        time we needn't update and just test for old json files running 
+        ``coala_html.main`` with noupdate. In the process we mock the
+        expensive http calls and instantiating the server by mocking the 
+        corresponding method calls.
```
Comment (Uran198) [inline tests/coalaHTMLTest.py]: Make it shorter, like:
```
Test that when JSON configs are already generated, coala-html with 'noupdate' option will run successfully and not update existing configuration.
```
Something like that. And change the name of the test, so it'll be closer to what behavior you're actually testing.

**Comment by Uran198** [comment]
ack 4aeb5df25de7b83d5769142be5edc3ab5ddd4e2c

**Comment by Uran198** [comment]
ack 5996cd677c72febaab5222ee1e595a15cca43395

**Comment by Uran198** [comment]
ack 5996cd677c72febaab5222ee1e595a15cca43395

File: coalahtml/coala_html.py (hunk 7 above)
Comment (tushar-rishav) [inline coalahtml/coala_html.py]: So I can think of something like:
* Check if the directories differ - True or False
* Have a `--reload` cli argument that if passed, would update the app provided directories differ  otherwise keep the files intact.
@Uran198  Seems sane?

Comment (Uran198) [inline coalahtml/coala_html.py]: LGTM. Only, it's better probably to check for inclusion, i. e. check that you have all you need. then it will not be big deal if user will have some custom files in there.

File: coalahtml/helper.py [hunk 14]
```diff
@@ -7,6 +8,43 @@
...
+    """
+    Compare the content of dir1 and dir2. In contrast with filecmp.dircmp,
+    this subclass compares the content of files with the same path.
+    """
+
+    def phase3(self):
+        """
+        Find out differences between common files.
+        Ensure we are using content comparison with shallow=False.
+        Ref: https://hg.python.org/cpython/file/2.7/Lib/filecmp.py#l174
```
Comment (Uran198) [inline coalahtml/helper.py]: Referencing to Python 2.7 doesn't seem right, cause you're not using that version. Also, check if this doesn't already exist in Python 3

File: coalahtml/helper.py [hunk 15]
```diff
@@ -95,4 +133,11 @@ def get_args():
                             metavar='BOOL',
                             default=False,
                             help='Do not launch webserver.')
+    arg_parser.add_argument('--noreload',
+                            nargs='?',
+                            const=True,
+                            metavar='BOOL',
+                            default=False,
+                            help='Do not update the local copy of app with'
+                                 'latest changes.')
```
Comment (Uran198) [inline coalahtml/helper.py]: missed space here. Should be
```
' latest changes.')
```

Comment (Uran198) [inline coalahtml/helper.py]: Also what about mentioning when it can be useful here? Not a lot, maybe some small sentence.

File: tests/coalaHTMLTest.py [hunk 16]
```diff
@@ -32,6 +32,33 @@ def tearDown(self):
         os.chdir(os.path.dirname(os.path.dirname(__file__)))
         shutil.rmtree('coalahtmlapp', ignore_errors=True)
 
+    @patch('coalahtml.coala_html.copy_files')
+    def test_reload(self, mock_copy_files):
+        """
+        With `--noreload` passed as the local copy should not be updated with
+        main package. With `--noreload` the local copy should remain intact.
```
Comment (Uran198) [inline tests/coalaHTMLTest.py]: These two sentences mean the same

**Comment by tushar-rishav** [comment]
reack 5996cd677c72febaab5222ee1e595a15cca43395 5206c9c4597e55534cbd72f71d61d022b0aacb79  4edb1b67eb987c94a638a63d4707f5d98ecc6da8 fef43776252ba43b4c6fe9c6f39a564d3aff61a3 ac86508dbbff5e6ca1cd0cff0ba17c0120c86715 41d27650ed0c24d9a9f4f3156f3087c2cfa25dcf

File: coalahtml/coala_html.py [hunk 17]
```diff
@@ -29,13 +29,34 @@
...
 
+def gen_file_set(dirpath):
+    """
+    Parse an output from `os.walk` and create a set of its contents.
+    :param dirpath:
+        Source directory.
+    :return:
+        A set of directory contents traversed recursively.
+    """
+    dir_content = list(os.walk(os.path.abspath(dirpath)))
```
Comment (Uran198) [inline coalahtml/coala_html.py]: No need to convert to list, just wasting memory

File: coalahtml/coala_html.py [hunk 18]
```diff
@@ -29,13 +29,34 @@
...
+    :param dirpath:
+        Source directory.
+    :return:
+        A set of directory contents traversed recursively.
+    """
+    dir_content = os.walk(os.path.abspath(dirpath))
+    content_set = set()
+    for dir_tuple in dir_content:
+        root, dir_list, file_list = dir_tuple
+        for file_name in [root] + dir_list + file_list:
```
Comment (Uran198) [inline coalahtml/coala_html.py]: I dislike this a little bit. You're creating a bit list, just for traversing it.
Doing stuff, like
```
dirpath = os.path.abspath(dirpath)
for root, dir_list, file_list in os.walk(dirpath):  # here I like to put os.walk here, as it will more apparent what you're doing, but not with abspath, cause it may get too complicated. Just personal preference, feel free to ignore.
        content_set.add(root)
        for file_name in itertools.chain(dir_list, file_list):  # this is not creating a huge list, so more effective on time and memory.
            content_set.add(file_name)
```
can be more effective.

File: coalahtml/coala_html.py [hunk 19]
```diff
@@ -29,13 +29,34 @@
...
     arg_parser = get_args()
     args = arg_parser.parse_args()
 
     dir_path = create_dir(os.path.abspath(args.dir))
 
-    if len(list(os.walk(dir_path))) > 0:  # pragma: no cover
+    main_package = gen_file_set(get_file(Constants.COALA_HTML_BASE))
+    local_package = gen_file_set(dir_path)
+    # Inclusion check
+    if (main_package & local_package) != main_package:  # pragma: no cover
```
Comment (Uran198) [inline coalahtml/coala_html.py]: use `issubset`: it'll be faster (probably), and the code will speak for itself.

**Comment by Uran198** [comment]
Too much backticks in aeb967c commit message, around `os.walk`

File: coalahtml/coala_html.py (hunk 18 above)
Comment (tushar-rishav) [inline coalahtml/coala_html.py]: Seems like I don't need `os.abspath` as the input is already absolute paths. (from `get_file` and `create_dir`) :)

File: coalahtml/coala_html.py (hunk 19 above)
Comment (tushar-rishav) [inline coalahtml/coala_html.py]: Python! :heart:

File: coalahtml/coala_html.py [hunk 20]
```diff
@@ -29,13 +30,33 @@
...
 def main():
     arg_parser = get_args()
     args = arg_parser.parse_args()
 
     dir_path = create_dir(os.path.abspath(args.dir))
 
-    if len(list(os.walk(dir_path))) > 0:  # pragma: no cover
+    main_package = gen_file_set(get_file(Constants.COALA_HTML_BASE))
+    local_package = gen_file_set(dir_path)
+    # Inclusion check
```
Comment (Uran198) [inline coalahtml/coala_html.py]: No need in this comments. Also, in general, comments should explain not what are you doing, but rather, why are you doing, what you're doing. What are you doing, should be clear from the code itself.

**Comment by Uran198** [comment]
That small change and we can merge.

**Comment by Uran198** [comment]
ack 232778be22f60bf4c67d2e0decaa1065860c852d

**Comment by Uran198** [comment]
ack 232778be22f60bf4c67d2e0decaa1065860c852d

**Comment by Uran198** [comment]
reack 232778be22f60bf4c67d2e0decaa1065860c852d

**Comment by Uran198** [comment]
@rultor merge

**Comment by tushar-rishav** [comment]
reack https://github.com/coala-analyzer/coala-html/commit/232778be22f60bf4c67d2e0decaa1065860c852d

**Comment by tushar-rishav** [comment]
@rultor merge
